### ISSUE TITLE
More diverse training

### ISSUE THREAD

**Issue by lambda** [gcp/leela-zero#1296]
In #1261 @gcp writes:

> The quality of play deteriorates rapidly with increasing handicap ("all moves lose no matter what I play?!") which is why I'm personally only interested in even games.

I'm not sure if this has already been considered, but I'm wondering about the value of training Leela Zero in more diverse game conditions to reduce this kind of brittleness.

For instance, playing the current generation against older nets with a handicap, to train for playing against someone with handicap stones. Giving one instance a severe time restriction, or no playout, and handicap stones, against one with no such restriction, to train for playing with handicap stones. "Blitz" games just training initial judgements without rollouts. Playing for margin of victory rather than simply to win, like bang-naegi, for training some more "fighting to the end" style. Not sure if it would be possible, due to issues with having to map the different board sizes to the same network, but playing on different board sizes (both smaller and larger). And maybe even playing against special opponents, like one that plays mirror go until the network computes a move of high enough value to be worth breaking the mirroring, or one that includes an explicit ladder engine, to help train better for these kinds of situations.

I could imagine a couple different ways to incorporate this. One would be to just include this in normal training, as a certain percentage of games. For some of these, you would need to include an input indicating which variant was in play at a given time, and would probably want to limit this to a small fraction of games played.

Another would be to fork LZ into two, one of which kept training like normal, and one which trained this way; then after some number of training rounds, start playing them against each other. Or you might even be able to fork into multiple, with different percentage of games played with each variant, and after they had each trained in each different "school" for several rounds, play against each other, select the strongest, and train them based on each other.

Anyhow, I feel like in order to reduce a little bit of the fragility of encountering situations which aren't encountered in pure self-play of standard 19x19 games, like handicap games, ladders, mirror go, etc, you could take some training variants that are generally played by humans, and have them included in the training.

**Comment by jkiliani** [comment]
Setting move selection temperature to 1 for the whole game instead of only the first 30 moves would have a significant effect in the direction you describe. Currently, LZ always plays the strongest move after move 30, which causes the value head to be rather pessimistic about the ability to reverse a disadvantage, and overly optimistic when ahead.

**Comment by gcp** [comment]
Not in scope for this project, unproven, complicated method, etc.

**Comment by Eddh** [comment]
Ideally i think the temperature should start high and then do it like in annealing : gradually decrease until 0.  It's a classic use of annealing : explore the space of moves first, then converge. Any thoughts of wether it would be doable for this run or another run ?

**Comment by jkiliani** [comment]
Temperature annealing is what Leela Chess uses for match games now to maintain game diversity. There is no consensus on using it for training games however, which could potentially work in my opinion if the annealing schedule is picked very carefully (i.e. really slow annealing).

For Leela Zero, it would be less needed for game diversity because the search space is much bigger and LZ has symmetries while Leela chess does not. I think it might help but obviously cannot prove that. I just have a slight aesthetic problem with Deepmind's switch at move 30, but in practise it seems to work well enough...

**Comment by Eddh** [comment]
No I don't mean temperature annealing inside a game, but with the whole training run, similar with learning rate annealing.

**Comment by jkiliani** [comment]
You're right about the playout difference, though I think this is less of a fundamental difference than an optimized parameter.

For Leela Zero, I would find it really interesting to simultaneously try
1. Always-promote with a window size around 500k
2. Proportional move selection for the whole game (i.e. temperature=1)
3. Visits reduced substantially (I'd say 1000, +1 for training data compression ratio)

But the project should progress just fine whether or not any of this ends up being tried. I suppose Leela Chess can now be viewed as a reference experiement that validates the AlphaZero procedure.

**Comment by jkiliani** [comment]
I just played Leela Zero (using Lizzie) at H4 and H3 and was able to beat it both times, due to its extremely conservative, territorial play (i.e. doing 3-3 invasions against handicap), me just being 1 dan. Against Leela 11, I have a lot more difficulty winning with such a handicap. I think that play style with handicap would really benefit a lot from proportional move selection in training games.

**Comment by Eddh** [comment]
No the problems with handicaps is not the playstyle. That's a symptom of the problem, which is that with high handicaps, the winrated are all below 5%. This makes it less able to differentiate between moves, and on top of that search parameters like puct and fpu are tuned for general winrate, which probably mean they are not optimal at these winrates, but rather for winrates where there is a decent chance of winning.

**Comment by alreadydone** [comment]
Comments welcome for my PR #1435
(not retraining the NN but modify the tree search for handicap play; implements my idea above)
